src/test/FreeC/IR/InliningTests.hs

```diff
@@ -0,0 +1,90 @@
+
+import           Test.Hspec
+
+
+import           FreeC.IR.Inlining              ( inlineExpr
+                                                , inlineFuncDecls
+                                                )
+import           FreeC.Monad.Class.Testable     ( shouldSucceedWith )
+import           FreeC.Test.Expectations
+import           FreeC.Test.Environment
+import           FreeC.Test.Parser              ( parseTestFuncDecl
+                                                , parseTestExpr
+                                                )
+
+-- | Test group for inlineExpr and inlineFuncDecls
+testInlining :: Spec
+testInlining = describe "FreeC.IR.Inlining" $ do
+  context "inlineExpr" $ do
+    it "inlines a simple function" $ shouldSucceedWith $ do
+      _        <- defineTestTypeCon "Integer" 0
+      _        <- defineTestFunc "g" 1 "forall a. a -> a"
+      fun      <- parseTestFuncDecl "f @a (x :: a) :: a = g @a x"
+      expr     <- parseTestExpr "f @Integer 42"
+      res      <- inlineExpr [fun] expr
+      expected <- parseTestExpr "g @Integer 42"
+      return (res `shouldBeSimilarTo` expected)
+    it "inlines transitive" $ shouldSucceedWith $ do
+      _        <- defineTestTypeCon "Integer" 0
+      _        <- defineTestFunc "h" 1 "forall a. a -> a"
+      funG     <- parseTestFuncDecl "g @a (x :: a) :: a = h @a x"
+      funF     <- parseTestFuncDecl "f @a (x :: a) :: a = g @a x "
+      expr     <- parseTestExpr "f @Integer 42"
+      res      <- inlineExpr [funF, funG] expr
+      expected <- parseTestExpr "h @Integer 42"
+      return (res `shouldBeSimilarTo` expected)
+    it "inlines in a function application" $ shouldSucceedWith $ do
+      _        <- defineTestTypeCon "Integer" 0
+      _        <- defineTestFunc "h" 1 "forall a. a -> a"
+      _        <- defineTestFunc "g" 1 "forall a. a -> a"
+      fun      <- parseTestFuncDecl "f @a (x :: a) :: a = h @a x"
+      expr     <- parseTestExpr "g @Integer (f @Integer 42)"
+      res      <- inlineExpr [fun] expr
+      expected <- parseTestExpr "g @Integer (h @Integer 42)"
+      return (res `shouldBeSimilarTo` expected)
+    it "inlines in an if expression" $ shouldSucceedWith $ do
+      _    <- defineTestTypeCon "Bool" 0
+      _    <- defineTestTypeCon "Integer" 0
+      fun  <- parseTestFuncDecl "f @a (x :: a) :: a = x"
+      expr <- parseTestExpr
+        "if (f @Bool True) then f @Integer 42 else f @Integer 24"
+      res      <- inlineExpr [fun] expr
+      expected <- parseTestExpr "if True then 42 else 24"
+      return (res `shouldBeSimilarTo` expected)
+  context "inlineFuncDecls" $ do
+    it "inlines a recursive function correctly" $ shouldSucceedWith $ do
+      _   <- defineTestTypeCon "Peano" 0
+      _   <- defineTestCon "Zero" 0 "Peano"
+      _   <- defineTestCon "Succ" 1 "Peano -> Peano"
+      add <-
+        parseTestFuncDecl
+          "add (m :: Peano) (n :: Peano) :: Peano = case m of { Zero -> n ; Succ m' -> Succ (add' m n) }"
```

Comment:
This line looks like it exceeds the 80 character line length limit. Brittany does not break up long string automatically. You have to insert `++` manually.

```suggestion
      add <-
        parseTestFuncDecl
        $  "add (m :: Peano) (n :: Peano) :: Peano = case m of {"
        ++ "   Zero -> n;"
        ++ "   Succ m' -> Succ (add' m n)"
        ++ "}"
```

Also check remaining lines in this file.